### PR TITLE
Custom Redis lwm2mpeer lwm2midentity serializers/deserlializers

### PR DESCRIPTION
**Pull request by JaroslawLegierski**
Possibility to create custom redis lwm2mpeer lwm2midentity serializers/deserlializers dedicated for "new Identity concept" [issue 1436](https://github.com/eclipse/leshan/issues/1436) 
&
Correction 500 Internal Server Error for unsecure devices after IP adress change

### CODE REVIEW

File: leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/send/SendResource.java [hunk 1]
```diff
@@ -65,6 +65,11 @@ public void handlePOST(CoapExchange exchange) {
             return;
         }
 
+        // if clientProfile exists but registration cannot be found because of device ip address change
+        if (clientProfile.getRegistration() == null) {
+            exchange.respond(ResponseCode.NOT_FOUND, "no registration found");
+            return;
+        }
```
Comment (sbernard31) [inline leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/send/SendResource.java:72]: I'm not sure I get this part of code ?

It seems you try to handle when there is no registration for a given device identity.
It this case, should we rather change `DefaultClientProfileProvider` behavior ? it should maybe return `null` if we didn't find any registration ?

File: leshan-server-demo/src/main/java/org/eclipse/leshan/server/demo/LeshanServerDemo.java [hunk 2]
```diff
@@ -158,7 +160,11 @@ public static LeshanServer createLeshanServer(LeshanServerDemoCLI cli) throws Ex
         } else {
             // use Redis Store
             securityStore = new RedisSecurityStore(cli.main.redis);
-            builder.setRegistrationStore(new RedisRegistrationStore(cli.main.redis));
+            RedisRegistrationStore.Builder redisregistrationstorebuilder = new RedisRegistrationStore.Builder(
+                    cli.main.redis);
+            redisregistrationstorebuilder.setIdentitySerDes(new LwM2mIdentitySerDes());
+            redisregistrationstorebuilder.setPeerSerDes(new LwM2mPeerSerDes());
+            builder.setRegistrationStore(redisregistrationstorebuilder.build());
```
Comment (sbernard31) [inline leshan-server-demo/src/main/java/org/eclipse/leshan/server/demo/LeshanServerDemo.java:167]: This seems not really needed right ?

File: leshan-server-redis/src/main/java/org/eclipse/leshan/server/redis/RedisRegistrationStore.java [hunk 3]
```diff
@@ -942,6 +948,24 @@ public Builder setRegistrationSerDes(RegistrationSerDes registrationSerDes) {
             return this;
         }
 
+        /**
+         * Set {@link LwM2mIdentitySerDes} instance used to LwM2mIdentity serialize/de-serialize {@link LwM2mIdentity}
```
Comment (sbernard31) [inline leshan-server-redis/src/main/java/org/eclipse/leshan/server/redis/RedisRegistrationStore.java:952]: > Set {@link LwM2mIdentitySerDes} instance used to ~LwM2mIdentity~ serialize/de-serialize {@link LwM2mIdentity}

Maybe `LwM2mIdentity` is not intended ?

File: leshan-server-redis/src/main/java/org/eclipse/leshan/server/redis/RedisRegistrationStore.java [hunk 4]
```diff
@@ -942,6 +948,24 @@ public Builder setRegistrationSerDes(RegistrationSerDes registrationSerDes) {
...
+         * to/from this store.
+         */
+        public Builder setIdentitySerDes(LwM2mIdentitySerDes identitySerDes) {
+            this.identitySerDes = identitySerDes;
+            return this;
+        }
+
+        /**
+         * Set {@link LwM2mPeerSerDes} instance used to LwM2mIdentity serialize/de-serialize {@link LwM2mPeer} to/from
```
Comment (sbernard31) [inline leshan-server-redis/src/main/java/org/eclipse/leshan/server/redis/RedisRegistrationStore.java:961]: > Set {@link LwM2mIdentitySerDes} instance used to ~LwM2mIdentity~ serialize/de-serialize {@link LwM2mIdentity}

Maybe `LwM2mIdentity` is not intended  too ?

File: leshan-server-redis/src/test/java/org/eclipse/leshan/server/redis/serialization/RegistrationSerDesTest.java [hunk 5]
```diff
@@ -39,7 +39,7 @@
 
 public class RegistrationSerDesTest {
 
-    private final RegistrationSerDes registrationSerDes = new RegistrationSerDes();
+    private final RegistrationSerDes registrationSerDes = new RegistrationSerDes(new LwM2mPeerSerDes());
```
Comment (sbernard31) [inline leshan-server-redis/src/test/java/org/eclipse/leshan/server/redis/serialization/RegistrationSerDesTest.java:42]: maybe instead we add a default `new RegistrationSerDes() `constructor ?

File: leshan-server-demo/src/main/java/org/eclipse/leshan/server/demo/LeshanServerDemo.java (hunk 2 above)
Comment (JaroslawLegierski) [reply leshan-server-demo/src/main/java/org/eclipse/leshan/server/demo/LeshanServerDemo.java]: You're right - it's part of my PoC - already fixed in the latest commit

File: leshan-server-redis/src/main/java/org/eclipse/leshan/server/redis/RedisRegistrationStore.java (hunk 4 above)
Comment (JaroslawLegierski) [reply leshan-server-redis/src/main/java/org/eclipse/leshan/server/redis/RedisRegistrationStore.java]: This is also corrected

File: leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/send/SendResource.java (hunk 1 above)
Comment (sbernard31) [reply leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/send/SendResource.java:72]: yep except that maybe we can get the model only if registration is not null ?

Comment (JaroslawLegierski) [reply leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/send/SendResource.java]: OK - corrected in [this](https://github.com/JaroslawLegierski/leshan/commit/9496ef01fbd19c296d67b98f9f6ad6f8f6f4531a) commit

**Pull request closed without merging**